docs/config-validation.md

```diff
@@ -0,0 +1,191 @@
+is the same as saying a struct and all of its fields are optional. We use this
+tag for many config duration and password file struct valued fields which are
+optional in some configs but required in others.
+
+### `structonly`
+
+Very similar to `omitempty`, but used to control whether or not the validator
+will run validations present for the fields of a nested struct. When a nested
+struct is tagged with `structonly` the validations for its fields will only run
+when the condition of a `required_` (`with`, `with_all`, `unless`, etc.) tag is
+met.
+
+### `nostructlevel`
+
+The same as `structonly`, but it will never run the validations on the nested
+struct, even if one of the `required_` conditions is met. When this tag is
+present, the validator will only validate the value of the struct itself is
+non-nil.
+
+### `min=1`, `gte=1`, `gt=0`
+
+These validate that the value is greater than zero. On its face you might assume
+that these would only validate integers, but when specified on a slice or map it
+will validate that the length of the slice or map is greater than zero.
+
+Note: we encourage the use of `min=1` or `gte=1` over `gt=0` because it's more
+readable.
+
+For instance, the following would be valid config for a slice valued field
+tagged with `required`.
+```json
+{
+  "foo": [],
+}
+```
+
+But, only the following would be valid config for a slice valued field tagged
+with `min=1`.
+```json
+{
+  "foo": ["bar", ...],
+}
+```
+
+### `len`
+
+Same as `eq` (equal to) but can also be used to validate the length of the
+strings.
+
+### `hostname_port`
+
+The
+[docs](https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-HostPort)
+for this tag are scant with detail, but it validates that the value is a valid
+RFC 1123 hostname and port. It is used to validate many of the the
+`ListenAddress` and `DebugAddr` fields of our components.
+
+#### Future Work
+
+This tag is compatible with IPv4 addresses, but not IPv6 addresses. We should
+consider fixing this in our fork of the validator.
+
+### `dive`
+This tag is used to validate the values of a slice or map. For instance, the
+following would be valid config for a slice valued field (`[]string`) tagged
+with `min=1,dive,oneof=bar baz`.
+
+```json
+{
+  "foo": [
+    "bar",
+    "baz",
+  ],
+}
+```
+
+Note that the `dive` tag introduces an order-dependence in writing tags: tags
+that come before `dive` apply to the current field, while tags that come after
+`dive` apply to the current field's child values. In the example above: `min=1`
+applies to the length of the slice (`[]string`), while `oneof=bar baz` applies
+to the value of each string in the slice.
+
+We can also use `dive` to validate the values of a map. For instance, the
+following would be valid config for a map valued field (`map[string]string`)
+tagged with `min=1,dive,oneof=bar baz`.
+
+```json
+{
+  "foo": {
+    "bar": "baz",
+    "baz": "bar"
```

Comment:
Use different keys here, so that it's doubly clear that the validation is applying to the values, not the keys.